File: rplugin/python3/curstr/action/source/vim/autoload_function.py
import os.path
import re
from glob import iglob
from typing import Tuple

from curstr.action.group import ActionGroup, FileDispatcher
from curstr.action.source import Source as Base


class Source(Base):

    DISPATCHER_CLASS = FileDispatcher

    def create(self) -> ActionGroup:
        cword = self._cursor.get_word(added_iskeyword="#")

        if "#" not in cword:
            return self._dispatcher.nothing()

        splited = cword.split("#")
        paths = splited[:-1]
        target = "{}.vim".format(os.path.join("", *paths))
        for runtimepath in self._vim.options["runtimepath"].split(","):
            file_path = os.path.join(runtimepath, "autoload", target)
            if os.path.isfile(file_path):
                position = self.__search_position(cword, file_path)
                return self._dispatcher.dispatch(
                    (
                        (FileDispatcher.FilePosition, file_path, *position),
                        (FileDispatcher.File, file_path),
                    )
                )

        if not self.get_option("include_packpath"):
            return self._dispatcher.nothing()

        package = self._vim.call(
            "fnamemodify", self._vim.options["packpath"].split(",")[0], ":p"
        )
        pattern = os.path.expanduser(
            os.path.join(package, *["pack/*/opt/*/autoload", target])
        )
        for path in iglob(pattern, recursive=True):
            if os.path.isfile(path):
                position = self.__search_position(cword, path)
                return self._dispatcher.dispatch(
                    (
                        (FileDispatcher.FilePosition, path, *position),
                        (FileDispatcher.File, path),
                    )
                )

        return self._dispatcher.nothing()

    def __search_position(self, function_name: str, path: str) -> Tuple[int, int]:
        with open(path, "r") as f:
            line = f.readline()
            row = 1
            while line:
                match = re.match(
                    "\\s*fu(nction)?!?\\s*(?P<name>{})\\(".format(function_name), line
                )
                if match is not None:
                    return (row, match.start("name") + 1)
                line = f.readline()
                row += 1

        return (0, 0)

    def get_options(self):
        return {"filetyps": ["vim"], "include_packpath": False}

File: lua/curstr/action_source/vim/autoload_function.lua
local M = {}

function M.create(ctx)
  local cword = require("curstr.lib.cursor").word("#")
  if cword:find("#") == nil then
    return nil
  end

  local path_parts = vim.split(cword, "#", { plain = true })
  table.remove(path_parts, #path_parts)
  local target = ("%s.vim"):format(table.concat(path_parts, "/"))
  local paths = vim.api.nvim_get_runtime_file("autoload/" .. target, true)
  for _, path in ipairs(paths) do
    if require("curstr.lib.file").readable(path) then
      local position = M._search(cword, path)
      return "file", { path = path, position = position }
    end
  end

  if not ctx.opts.include_packpath then
    return nil
  end

  local pack_path = vim.split(vim.o.packpath, ",", { plain = true })[1]
  local package = vim.fn.fnamemodify(pack_path, ":p")
  local pattern = vim.fs.joinpath(package, "pack/*/opt/*/autoload", target)
  for _, path in ipairs(vim.fn.glob(pattern, false, true)) do
    if require("curstr.lib.file").readable(path) then
      local position = M._search(cword, path)
      return "file", { path = path, position = position }
    end
  end

  return nil
end

function M._search(name, path)
  local f = io.open(path, "r")
  assert(f, "cannot open file: " .. path)
  local row = 1
  local pattern = ("\\v\\s*fu(nction)?!?\\s*\\zs%s\\ze\\("):format(name)
  local regex = vim.regex(pattern)
  for line in f:lines() do
    local s = regex:match_str(line)
    if s ~= nil then
      f:close()
      return { row, s + 1 }
    end
    row = row + 1
  end
  f:close()
  return nil
end

M.opts = {
  include_packpath = false,
}

M.filetypes = { "vim" }

M.description = [[searches a vim autoload function]]

return M
